File: client/verification/verification.go
package verification

import (
	"encoding/binary"
	"errors"
	"slices"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

type Request []byte

type Response []byte

// isStaticType checks whether bytes that represent abi.encoded request or response that encodes an instance of static type.
// abi.encode(X) = enc((X)) of X of type T is encoding of tuple (X) of type (T). By specification, enc((X)) = head(X)tail(X).
// If T is static, head(X) = enc(X) and tail(X) is empty. If T is dynamic, head(X) = bytes32(len(head(X))) = bytes32(32) and tail = enc(X).
// See https://docs.soliditylang.org/en/latest/abi-spec.html for detailed specification.
func isStaticType(bytes []byte) (bool, error) {

	if len(bytes) < 96 {
		return false, errors.New("bytes are to short")
	}

	first32Bytes := [32]byte(bytes[:32])

	d := [32]byte{}

	d[31] = byte(32)

	return d != first32Bytes, nil

}

// AttestationType returns the attestation type of the request (the first 32 bytes).
func (r Request) AttestationType() ([32]byte, error) {

	res := [32]byte{}

	if len(r) < 96 {
		return res, errors.New("request is to short")
	}

	copy(res[:], r[0:32])

	return res, nil

}

// Source returns the source of the request (the second 32 bytes).
func (r Request) Source() ([32]byte, error) {

	res := [32]byte{}

	if len(r) < 96 {
		return res, errors.New("request is to short")
	}

	copy(res[:], r[32:64])

	return res, nil
}

// AttestationTypeAndSource returns byte encoded AttestationType and Source (the first 64 bytes).
func (r Request) AttestationTypeAndSource() ([64]byte, error) {

	res := [64]byte{}

	if len(r) < 96 {
		return [64]byte{}, errors.New("request is to short")
	}

	copy(res[:], r[0:64])

	return res, nil
}

// Mic returns Message Integrity code of the request (the third 32 bytes).
func (r Request) Mic() (common.Hash, error) {

	if len(r) < 96 {
		return common.Hash{}, errors.New("request is to short")
	}

	mic := common.Hash{}

	mic.SetBytes(r[64:96])

	return mic, nil
}

// ComputeMic computes Mic from the response.
// Mic is the hash of the response with roundID set to 0.
func (r Response) ComputeMicMaybe() (common.Hash, error) {

	static, err := isStaticType(r)

	if err != nil {
		return common.Hash{}, err
	}

	// roundId is encoded in the third 32bytes slot
	roundIdStartByte := 64
	roundIdEndByte := 96
	commonFieldsLength := 128

	// if Response is encoded dynamic struct the first 32 bytes are bytes32(32)
	if !static {
		roundIdStartByte += 32
		roundIdEndByte += 32
		commonFieldsLength += 32
	}

	if len(r) < commonFieldsLength {
		return common.Hash{}, errors.New("response is to short")
	}

	// store roundId
	d := make([]byte, 32)
	roundIdBytes := r[roundIdStartByte:roundIdEndByte]
	copy(d, roundIdBytes)

	// restore roundId at the end
	defer copy(roundIdBytes, d)

	// set roundId to zero
	zero32bytes := make([]byte, 32)
	copy(roundIdBytes, zero32bytes)

	mic := crypto.Keccak256Hash(r)

	return mic, nil

}

func (r Response) ComputeMic(args abi.Arguments) (common.Hash, error) {

	decoded, err := args.Unpack(r)

	if err != nil {
		return common.Hash{}, err
	}

	stringArgument := abi.Argument{}

	stringArgument.Name = "string"
	stringArgument.Indexed = false

	stringArgument.Type, err = abi.NewType("string", "string", []abi.ArgumentMarshaling{})

	if err != nil {
		return common.Hash{}, err
	}

	args = append(args, stringArgument)

	withSalt, err := args.Pack(decoded[0], "Flare")

	if err != nil {
		return common.Hash{}, err
	}

	mic := crypto.Keccak256Hash(withSalt)

	return mic, nil

}

// AddRound sets the roundId in the response (third 32 bytes).
func (r Response) AddRound(roundId uint64) (Response, error) {

	static, err := isStaticType(r)

	if err != nil {
		return Response{}, err
	}

	// roundId is encoded in the third slot
	roundIdStartByte := 64
	roundIdEndByte := 96
	commonFieldsLength := 128

	// if Response is encoded dynamic struct the first 32 bytes are bytes32(32)
	if !static {
		roundIdStartByte += 32
		roundIdEndByte += 32
		commonFieldsLength += 32
	}

	if len(r) < commonFieldsLength {
		return Response{}, errors.New("response is to short")
	}

	roundIdEncoded := make([]byte, 0)

	binary.BigEndian.AppendUint64(roundIdEncoded, roundId)

	roundIdSlot := make([]byte, 32-len(roundIdEncoded))

	roundIdSlot = append(roundIdSlot, roundIdEncoded...)

	r = slices.Replace(r, roundIdStartByte, roundIdEndByte, roundIdSlot...)

	return r, nil

}

// Hash computes hash of the response.
func (r Response) Hash() (common.Hash, error) {
	if len(r) < 128 {
		return common.Hash{}, errors.New("response is to short")
	}

	hash := crypto.Keccak256Hash(r)

	return hash, nil
}

File: src/attestation/verification.py
from typing import NewType, List, Tuple
from eth_abi import encode as abi_encode, decode as abi_decode
from Crypto.Hash import keccak
from src.logger import log

# Type aliases to mirror the Go implementation
Request = NewType("Request", bytes)
Response = NewType("Response", bytes)

def get_attestation_type(r: Request) -> bytes:
    """Extracts the attestation type from the request (first 32 bytes)."""
    if len(r) < 96:
        log.error("Request is too short to extract attestation type.")
        raise ValueError("Request is too short")
    return r[0:32]

def get_source(r: Request) -> bytes:
    """Extracts the source from the request (second 32 bytes)."""
    if len(r) < 96:
        log.error("Request is too short to extract source.")
        raise ValueError("Request is too short")
    return r[32:64]

def get_mic(r: Request) -> bytes:
    """Extracts the Message Integrity Code (MIC) from the request (third 32 bytes)."""
    if len(r) < 96:
        log.error("Request is too short to extract MIC.")
        raise ValueError("Request is too short")
    return r[64:96]

def valid_lut(lut: int, lut_limit: int, round_start: int) -> bool:
    """
    Safely checks whether round_start - lut < lut_limit.
    This function uses integer arithmetic and is a direct port of the Go logic.
    """
    log.debug(f"Validating LUT: lut={lut}, lut_limit={lut_limit}, round_start={round_start}")
    if lut > round_start:
        log.debug("LUT is in the future, invalid.")
        return False

    lhs = round_start - lut
    is_valid = lhs < lut_limit
    log.debug(f"round_start - lut = {lhs}. Is valid: {is_valid}")
    return is_valid

def is_static_type(encoded_bytes: bytes) -> bool:
    """
    Checks if ABI-encoded bytes represent a static or dynamic type.
    If the first 32 bytes represent the offset to the data (i.e., 32),
    it's a dynamic type. Otherwise, it's static.
    """
    if len(encoded_bytes) < 32:
        log.error("Encoded bytes are too short to determine type.")
        raise ValueError("Encoded bytes are too short to determine type.")

    offset = int.from_bytes(encoded_bytes[0:32], 'big')
    is_dynamic = offset == 32
    log.debug(f"Offset is {offset}. Is dynamic type: {is_dynamic}")
    return not is_dynamic

def get_lut(response: Response) -> int:
    """
    Returns the LUT (Latest Used Timestamp) from the response.
    LUT is assumed to be a uint64 in the fourth 32-byte slot.
    """
    static = is_static_type(response)
    log.debug(f"Response is a static type: {static}")

    lut_start = 3 * 32
    lut_end = 4 * 32

    if not static:
        lut_start += 32
        lut_end += 32

    if len(response) < lut_end:
        log.error("Response is too short to contain LUT.")
        raise ValueError("Response is too short to contain LUT")

    lut_bytes = response[lut_start:lut_end]
    lut = int.from_bytes(lut_bytes, 'big')
    log.debug(f"Extracted LUT: {lut}")
    return lut

def add_round(response: Response, round_id: int) -> Response:
    """
    Sets the round ID in the response (third 32-byte slot).
    Returns a new Response bytes object with the round ID inserted.
    """
    static = is_static_type(response)
    log.debug(f"Response is a static type: {static}")

    round_id_start = 2 * 32
    round_id_end = 3 * 32

    if not static:
        round_id_start += 32
        round_id_end += 32

    if len(response) < round_id_end:
        log.error("Response is too short to add round ID.")
        raise ValueError("Response is too short to add round ID")

    round_id_bytes = round_id.to_bytes(32, 'big')
    log.debug(f"Inserting round ID {round_id} into response.")

    new_response_bytes = (
        response[:round_id_start] +
        round_id_bytes +
        response[round_id_end:]
    )

    return Response(new_response_bytes)

def get_hash(response: Response, round_id: int) -> bytes:
    """
    Computes the final hash of the response after adding the round ID.
    """
    log.debug(f"Computing hash for response with round ID {round_id}")
    response_with_round = add_round(response, round_id)

    hasher = keccak.new(digest_bits=256)
    hasher.update(response_with_round)
    final_hash = hasher.digest()
    log.debug(f"Computed final hash: {final_hash.hex()}")
    return final_hash

def compute_mic(response: Response, response_abi_types: List[str]) -> bytes:
    """
    Computes the Message Integrity Code (MIC) from the response.
    MIC is keccak256(abi.encode(abi.decode(response), "Flare")).
    """
    log.debug(f"Computing MIC for response with ABI types: {response_abi_types}")
    decoded_data = abi_decode(response_abi_types, response)
    log.debug(f"Decoded data: {decoded_data}")

    types_with_salt = response_abi_types + ['string']
    data_with_salt = list(decoded_data) + ["Flare"]
    log.debug(f"Data with salt: {data_with_salt}")

    encoded_with_salt = abi_encode(types_with_salt, data_with_salt)

    hasher = keccak.new(digest_bits=256)
    hasher.update(encoded_with_salt)
    mic = hasher.digest()
    log.debug(f"Computed MIC: {mic.hex()}")
    return mic